README.md

```diff
@@ -31,8 +45,161 @@ $ docker-compose up
+interview.
+
+### Programming tasks
+
+#### 1. General understanding
+
+Before the first task, ask the candidate to evaluate the code in the main
+server source code file. Invite them to ask you questions or otherwise
+clarify any confusing elements. Allow them to scan the framework documentation
+or explore the Docker file that creates the server if they express an interest.
+
+##### Evaluation criteria
+
+1. The candidate may not be proficient in either programming language
+    available, but this gives us the opportunity to assess how quickly they may
+    be able to acquire these skills based on their ability to translate their
+    existing knowledge.
+2. Do they understand the basics of HTTP requests and responses and map these
+   concepts correctly to the framework?
+3. Do they demonstrate methodical learning by structuring their questions
+   clearly and specifically?
+
+#### 2. Persist the key/values between process restarts
+
+If the server process crashes or is reloaded during code changes, the in-memory data
+structure is reinitialized. Have the candidate persist the keys/values in a way that
+makes them resilient to server restarts. This will likely involve writing to disk
+and reloading the data, but allow the candidate to suggest possible implementations.
+
+##### Evaluation criteria
+
+Does the candidate:
+
+1. Understand why the in-memory structure works, and why it is empty after a
+   restart?
+2. Understand basic I/O concepts?
+3. Discuss tradeoffs between the different options of when to save and load
+   data? Depending on their implementation, this could involve:
+   - Having two sources of truth in a file persisted data store and an in-memory store?
+   - Performance implications of reading the entire file on every request vs at start-up?
+4. Understand the life cycle of server: process re/start, request, response and where
+   in the framework they can respond to them?
+5. Handle empty cases (i.e., starting with no keys loaded)?
+6. Allow duplicate keys? The in-memory dictionary structure updates existing keys.
+
+#### 3. Fetch key by name
+
+Add a new endpoint that returns the value of the provided key.
+
+##### Evaluation criteria
+
+Does the candidate:
+
+1. Understand how to add a new route and function?
+2. Use sensible names for the path and the key?
+3. Demonstrate understanding of the methods for encoding the query in the
+   request? (i.e. query string param vs path argument vs form encoded POST)
+4. Handle common edge cases, like non-existing keys and null keys?
+5. Are they concerned about sanitizing user input?
+
+#### 4. Filter keys by prefix
+
+Add or extend an endpoint to return a list of all keys and values that match the
+provided prefix. For example, the prefix "na" would match "name" and "nationality".
+
+As an extension, have them sort the returned list in lexicographical key order.
+
+##### Evaluation criteria
+
+1. Is their method of filtering keys efficient?
```

Comment:
What were you thinking here? Keeping the data structure sorted in-memory? Only traversing the list once vs. all of the keys, then a lookup for each match?